trace-utils/src/msgpack_decoder/v04/decoder/span.rs

```diff
@@ -0,0 +1,159 @@
+use std::str::FromStr;
+
+/// Decodes a slice of bytes into a `Span` object.
+///
+/// # Arguments
+///
+/// * `buf` - A mutable reference to a slice of bytes containing the encoded data.
+///
+/// # Returns
+///
+/// * `Ok(Span)` - A decoded `Span` object if successful.
+/// * `Err(DecodeError)` - An error if the decoding process fails.
+///
+/// # Errors
+///
+/// This function will return an error if:
+/// - The map length cannot be read.
+/// - Any key or value cannot be decoded.
+#[inline]
+pub(crate) fn decode_span(buf: &mut &[u8]) -> Result<Span, DecodeError> {
+    let mut span = Span::default();
+    let span_size = rmp::decode::read_map_len(buf).map_err(|_| {
+        DecodeError::InvalidFormat("Unable to get map len for span size".to_owned())
+    })?;
+
+    for _ in 0..span_size {
+        fill_span(&mut span, buf)?;
+    }
+    Ok(span)
+}
+
+#[derive(Debug, PartialEq)]
+enum SpanKey {
+    Service,
+    Name,
+    Resource,
+    TraceId,
+    SpanId,
+    ParentId,
+    Start,
+    Duration,
+    Error,
+    Meta,
+    Metrics,
+    Type,
+    MetaStruct,
+    SpanLinks,
+}
+
+impl FromStr for SpanKey {
+    type Err = DecodeError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "service" => Ok(SpanKey::Service),
+            "name" => Ok(SpanKey::Name),
+            "resource" => Ok(SpanKey::Resource),
+            "trace_id" => Ok(SpanKey::TraceId),
+            "span_id" => Ok(SpanKey::SpanId),
+            "parent_id" => Ok(SpanKey::ParentId),
+            "start" => Ok(SpanKey::Start),
+            "duration" => Ok(SpanKey::Duration),
+            "error" => Ok(SpanKey::Error),
+            "meta" => Ok(SpanKey::Meta),
+            "metrics" => Ok(SpanKey::Metrics),
+            "type" => Ok(SpanKey::Type),
+            "meta_struct" => Ok(SpanKey::MetaStruct),
+            "span_links" => Ok(SpanKey::SpanLinks),
+            _ => Err(DecodeError::InvalidFormat(
+                format!("Invalid span key: {}", s).to_owned(),
+            )),
+        }
+    }
+}
+
+fn fill_span(span: &mut Span, buf: &mut &[u8]) -> Result<(), DecodeError> {
+    // field's key won't be held so no need to copy it in a buffer.
+    let (key, value) = read_string_ref(buf)?;
+
+    // Go to the value
+    *buf = value;
+
+    let key = key.parse::<SpanKey>()?;
+
+    match key {
+        SpanKey::Service => {
+            let (value, next) = read_string_ref(buf)?;
+            span.service = String::from_str(value).unwrap();
+            *buf = next;
+        }
+        SpanKey::Name => {
+            let (value, next) = read_string_ref(buf)?;
+            span.name = String::from_str(value).unwrap();
+            *buf = next;
+        }
+        SpanKey::Resource => {
+            let (value, next) = read_string_ref(buf)?;
+            span.resource = String::from_str(value).unwrap();
+            *buf = next;
+        }
+        SpanKey::TraceId => span.trace_id = read_number(buf)?.try_into()?,
```

Comment:
Maybe not necessary for this PR, but I think it would be good if the `read_number`, `read_string`, etc. had versions where the _field name_ was provided so the error messages contained what field we were trying to decode when things went wrong.

Reply:
I agree. I'll create a ticket to address this later.